bootloader/core/util.c

```diff
@@ -78,3 +78,43 @@ uint16_t compute_checksum(const uint8_t * begin, const uint8_t * end) {
     }
     return (~computed_checksum + 1) & 0xFFFF;
 }
+
+
+
+/**
+ * Iterate a byte buffer creating double word values and addresses.
+ *
+ * Writes to G4 and L5 flash happen in 64 bit data increments.
+ * @param address The starting flash address
+ * @param buffer A byte buffer
+ * @param length Length in bytes of the buffer.
+ * @param callback A callback that will be called for each 64 bit address
+ * @return True on success
+ */
+bool dword_address_iter(uint32_t address, const uint8_t * buffer, uint8_t length, address_iter_callback callback) {
+    if (!buffer || length <= 0) {
+        return true;
+    }
+    uint64_t double_word = 0;
+    for (int i = 0; i < length; i++) {
+        uint8_t double_word_index = i % sizeof(uint64_t);
+        if (i != 0 && double_word_index == 0) {
+            // We have a complete double word. Call the callback.
+            if (!callback(address, double_word)) {
+                return false;
+            }
+            // Clear word.
+            double_word = 0;
+            // Increment address
+            address += sizeof(uint64_t);
+        }
+        // The number of bytes to shift.
+        int shift_bytes = sizeof(uint64_t) - (double_word_index + 1);
+        int shift_amount = shift_bytes * 8;
+        // Or the shifted byte into the double word
+        double_word |= ((uint64_t)buffer[i] << shift_amount);
```

Comment:
Is this what is adding the word to the double word? 
If so, what is the bitwise OR doing?

Reply:
`double_word` is accumulating bytes until a full 8 bytes have been written into it. Shift amount goes from 7 to 0 repeatedly.